// @esri/arcgis-rest-items/dist/umd/items.umd.js
/* @preserve
* @esri/arcgis-rest-items - v1.19.0 - Apache-2.0
* Copyright (c) 2017-2019 Esri, Inc.
* Thu Mar 28 2019 12:28:54 GMT-0700 (Pacific Daylight Time)
*/
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('@esri/arcgis-rest-request')) :
    typeof define === 'function' && define.amd ? define(['exports', '@esri/arcgis-rest-request'], factory) :
    (factory((global.arcgisRest = global.arcgisRest || {}),global.arcgisRest));
}(this, (function (exports,arcgisRestRequest) { 'use strict';

    /*! *****************************************************************************
    Copyright (c) Microsoft Corporation. All rights reserved.
    Licensed under the Apache License, Version 2.0 (the "License"); you may not use
    this file except in compliance with the License. You may obtain a copy of the
    License at http://www.apache.org/licenses/LICENSE-2.0

    THIS CODE IS PROVIDED ON AN *AS IS* BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
    KIND, EITHER EXPRESS OR IMPLIED, INCLUDING WITHOUT LIMITATION ANY IMPLIED
    WARRANTIES OR CONDITIONS OF TITLE, FITNESS FOR A PARTICULAR PURPOSE,
    MERCHANTABLITY OR NON-INFRINGEMENT.

    See the Apache Version 2.0 License for specific language governing permissions
    and limitations under the License.
    ***************************************************************************** */

    var __assign = function() {
        __assign = Object.assign || function __assign(t) {
            for (var s, i = 1, n = arguments.length; i < n; i++) {
                s = arguments[i];
                for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p)) t[p] = s[p];
            }
            return t;
        };
        return __assign.apply(this, arguments);
    };

    /* Copyright (c) 2017-2018 Environmental Systems Research Institute, Inc.
     * Apache-2.0 */
    /**
     * Serialize an item and its data into a json format accepted by the Portal API for create and update operations
     *
     * @param item Item to be serialized
     * @returns a formatted json object to be sent to Portal
     */
    function serializeItem(item) {
        // create a clone so we're not messing with the original
        var clone = JSON.parse(JSON.stringify(item));
        // convert .data to .text
        if (clone.data) {
            clone.text = clone.data;
            delete clone.data;
        }
        return clone;
    }
    /**
     * requestOptions.owner is given priority, requestOptions.item.owner will be checked next. If neither are present, authentication.username will be assumed.
     */
    function determineOwner(requestOptions) {
        if (requestOptions.owner) {
            return requestOptions.owner;
        }
        if (requestOptions.item && requestOptions.item.owner) {
            return requestOptions.item.owner;
        }
        else {
            return requestOptions.authentication.username;
        }
    }

    /* Copyright (c) 2018 Environmental Systems Research Institute, Inc.
     * Apache-2.0 */
    /**
     * ```js
     * import { addItemJsonData } from '@esri/arcgis-rest-items';
     * //
     * addItemJsonData({
     *   id: '3ef',
     *   data: {}
     *   authentication
     * })
     *   .then(response)
     * ```
     * Send json to an item to be stored as the `/data` resource. See the [REST Documentation](https://developers.arcgis.com/rest/users-groups-and-items/update-item.htm) for more information.
     *
     * @param requestOptions - Options for the request
     * @returns A Promise that will resolve with an object reporting
     *        success/failure and echoing the item id.
     */
    function addItemJsonData(requestOptions) {
        var owner = determineOwner(requestOptions);
        var url = arcgisRestRequest.getPortalUrl(requestOptions) + "/content/users/" + owner + "/items/" + requestOptions.id + "/update";
        // Portal API requires that the 'data' be stringified and POSTed in
        // a `text` form field. It can also be sent with the `.create` call by sending
        // a `.data` property.
        requestOptions.params = __assign({ text: requestOptions.data }, requestOptions.params);
        return arcgisRestRequest.request(url, requestOptions);
    }
    /**
     * ```js
     * import { addItemData } from '@esri/arcgis-rest-items';
     * //
     * addItemData({
     *   id: '3ef',
     *   data: file,
     *   authentication
     * })
     *   .then(response)
     * ```
     * Send a file or blob to an item to be stored as the `/data` resource. See the [REST Documentation](https://developers.arcgis.com/rest/users-groups-and-items/update-item.htm) for more information.
     *
     * @param requestOptions - Options for the request
     * @returns A Promise that will resolve with an object reporting
     *        success/failure and echoing the item id.
     */
    function addItemData(requestOptions) {
        var owner = determineOwner(requestOptions);
        var url = arcgisRestRequest.getPortalUrl(requestOptions) + "/content/users/" + owner + "/items/" + requestOptions.id + "/update";
        // Portal API requires that the 'data' be POSTed in a `file` form field.
        requestOptions.params = __assign({ file: requestOptions.data }, requestOptions.params);
        return arcgisRestRequest.request(url, requestOptions);
    }
    /**
     * ```js
     * import { addItemRelationship } from '@esri/arcgis-rest-items';
     * //
     * addItemRelationship({
     *   originItemId: '3ef',
     *   destinationItemId: 'ae7',
     *   relationshipType: 'Service2Layer',
     *   authentication
     * })
     *   .then(response)
     * ```
     * Add a relationship between two items. See the [REST Documentation](https://developers.arcgis.com/rest/users-groups-and-items/add-relationship.htm) for more information.
     *
     * @param requestOptions - Options for the request
     * @returns A Promise to add item resources.
     */
    function addItemRelationship(requestOptions) {
        var owner = determineOwner(requestOptions);
        var url = arcgisRestRequest.getPortalUrl(requestOptions) + "/content/users/" + owner + "/addRelationship";
        var options = __assign({ params: {} }, requestOptions);
        arcgisRestRequest.appendCustomParams(requestOptions, options);
        return arcgisRestRequest.request(url, options);
    }
    /**
     * ```js
     * import { addItemResource } from '@esri/arcgis-rest-items';
     * //
     * addItemResource({
     *   id: '3ef',
     *   resource: file,
     *   name: 'bigkahuna.jpg',
     *   authentication
     * })
     *   .then(response)
     * ```
     * Add a resource associated with an item. See the [REST Documentation](https://developers.arcgis.com/rest/users-groups-and-items/add-resources.htm) for more information.
     *
     * @param requestOptions - Options for the request
     * @returns A Promise to add item resources.
     */
    function addItemResource(requestOptions) {
        var owner = determineOwner(requestOptions);
        var url = arcgisRestRequest.getPortalUrl(requestOptions) + "/content/users/" + owner + "/items/" + requestOptions.id + "/addResources";
        requestOptions.params = __assign({ file: requestOptions.resource, fileName: requestOptions.name, text: requestOptions.content, access: requestOptions.private ? "private" : "inherit" }, requestOptions.params);
        return arcgisRestRequest.request(url, requestOptions);
    }

    /* Copyright (c) 2018 Environmental Systems Research Institute, Inc.
     * Apache-2.0 */
    /**
     * ```js
     * import { createFolder } from '@esri/arcgis-rest-items';
     * //
     * createFolder({
     *   title: 'Map Collection',
     *   authentication: userSession
     * })
     *   .then(response)
     * ```
     * Create a folder. See the [REST Documentation](https://developers.arcgis.com/rest/users-groups-and-items/create-folder.htm) for more information.
     *
     * @param requestOptions - Options for the request
     * @returns A Promise that resolves with folder details once the folder has been created
     */
    function createFolder(requestOptions) {
        var owner = determineOwner(requestOptions);
        var baseUrl = arcgisRestRequest.getPortalUrl(requestOptions) + "/content/users/" + owner;
        var url = baseUrl + "/createFolder";
        requestOptions.params = __assign({ title: requestOptions.title }, requestOptions.params);
        return arcgisRestRequest.request(url, requestOptions);
    }
    /**
     * ```js
     * import { createItemInFolder } from '@esri/arcgis-rest-items';
     * //
     * createItem({
     *   item: {
     *     title: "The Amazing Voyage",
     *     type: "Web Map",
     *     data: {}
     *   },
     *   folder: 'underwater',
     *   authentication
     * })
     * ```
     * Create an item in a folder. See the [REST Documentation](https://developers.arcgis.com/rest/users-groups-and-items/add-item.htm) for more information.
     *
     * @param requestOptions = Options for the request
     */
    function createItemInFolder(requestOptions) {
        var owner = determineOwner(requestOptions);
        var baseUrl = arcgisRestRequest.getPortalUrl(requestOptions) + "/content/users/" + owner;
        var url = baseUrl + "/addItem";
        if (requestOptions.folder) {
            url = baseUrl + "/" + requestOptions.folder + "/addItem";
        }
        // serialize the item into something Portal will accept
        requestOptions.params = __assign({}, requestOptions.params, serializeItem(requestOptions.item));
        return arcgisRestRequest.request(url, requestOptions);
    }
    /**
     * ```js
     * import { createItem } from '@esri/arcgis-rest-items';
     * //
     * createItem({
     *   item: {
     *     title: "The Amazing Voyage",
     *     type: "Web Map",
     *     data: {}
     *   },
     *   authentication
     * })
     * ```
     * Create an Item in the user's root folder. See the [REST Documentation](https://developers.arcgis.com/rest/users-groups-and-items/add-item.htm) for more information.
     *
     * @param requestOptions - Options for the request
     * @returns A Promise that creates an item.
     */
    function createItem(requestOptions) {
        // delegate to createItemInFolder placing in the root of the filestore
        var options = __assign({ folder: null }, requestOptions);
        return createItemInFolder(options);
    }

    /* Copyright (c) 2018 Environmental Systems Research Institute, Inc.
     * Apache-2.0 */
    /**
     * ```
     * import { getItem } from "@esri/arcgis-rest-items";
     * //
     * getItem("ae7")
     *   .then(response);
     * // or
     * getItem("ae7", { authentication })
     *   .then(response)
     * ```
     * Get an item by id. See the [REST Documentation](https://developers.arcgis.com/rest/users-groups-and-items/item.htm) for more information.
     *
     * @param id - Item Id
     * @param requestOptions - Options for the request
     * @returns A Promise that will resolve with the data from the response.
     */
    function getItem(id, requestOptions) {
        var url = arcgisRestRequest.getPortalUrl(requestOptions) + "/content/items/" + id;
        // default to a GET request
        var options = __assign({ httpMethod: "GET" }, requestOptions);
        return arcgisRestRequest.request(url, options);
    }
    /**
     * ```
     * import { getItemData } from "@esri/arcgis-rest-items";
     * //
     * getItemData("ae7")
     *   .then(response)
     * // or
     * getItemData("ae7", { authentication })
     *   .then(response)
     * ```
     * Get the /data for an item. See the [REST Documentation](https://developers.arcgis.com/rest/users-groups-and-items/item-data.htm) for more information.
     * @param id - Item Id
     * @param requestOptions - Options for the request
     * @returns A Promise that will resolve with the json data for the item.
     */
    function getItemData(id, requestOptions) {
        var url = arcgisRestRequest.getPortalUrl(requestOptions) + "/content/items/" + id + "/data";
        // default to a GET request
        var options = __assign({ httpMethod: "GET", params: {} }, requestOptions);
        if (options.file) {
            options.params.f = null;
        }
        return arcgisRestRequest.request(url, options);
    }
    /**
     * ```
     * import { getRelatedItems } from "@esri/arcgis-rest-items";
     * //
     * getRelatedItems({
     *   id: "ae7",
     *   relationshipType: "Service2Layer" // or several ["Service2Layer", "Map2Area"]
     * })
     *   .then(response)
     * ```
     * Get the related items. See the [REST Documentation](https://developers.arcgis.com/rest/users-groups-and-items/related-items.htm) for more information.
     *
     * @param requestOptions - Options for the request
     * @returns A Promise to get some item resources.
     */
    function getRelatedItems(requestOptions) {
        var url = arcgisRestRequest.getPortalUrl(requestOptions) + "/content/items/" + requestOptions.id + "/relatedItems";
        var options = __assign({ httpMethod: "GET", params: {
                direction: requestOptions.direction
            } }, requestOptions);
        if (typeof requestOptions.relationshipType === "string") {
            options.params.relationshipType = requestOptions.relationshipType;
        }
        else {
            options.params.relationshipTypes = requestOptions.relationshipType;
        }
        delete options.direction;
        delete options.relationshipType;
        return arcgisRestRequest.request(url, options);
    }
    /**
     * Get the resources associated with an item
     *
     * @param requestOptions - Options for the request
     * @returns A Promise to get some item resources.
     */
    function getItemResources(requestOptions) {
        var url = arcgisRestRequest.getPortalUrl(requestOptions) + "/content/items/" + requestOptions.id + "/resources";
        // mix in user supplied params
        requestOptions.params = __assign({}, requestOptions.params, { num: 1000 });
        // at v2, the argument signature of this method should match getItemData() and getItemGroups() if requests can be made anonymously
        return arcgisRestRequest.request(url, requestOptions);
    }
    /**
     * ```js
     * import { getItemGroups } from "@esri/arcgis-rest-items";
     * //
     * getItemGroups("30e5fe3149c34df1ba922e6f5bbf808f")
     *   .then(response)
     * ```
     * Lists the groups of which the item is a part, only showing the groups that the calling user can access. See the [REST Documentation](https://developers.arcgis.com/rest/users-groups-and-items/groups.htm) for more information.
     *
     * @param id - The Id of the item to query group association for.
     * @param requestOptions - Options for the request
     * @returns A Promise to get some item groups.
     */
    function getItemGroups(id, requestOptions) {
        var url = arcgisRestRequest.getPortalUrl(requestOptions) + "/content/items/" + id + "/groups";
        return arcgisRestRequest.request(url, requestOptions);
    }

    /* Copyright (c) 2018 Environmental Systems Research Institute, Inc.
     * Apache-2.0 */
    /**
     * Protect an item. See the [REST Documentation](https://developers.arcgis.com/rest/users-groups-and-items/protect.htm) for more information.
     *
     * @param requestOptions - Options for the request
     * @returns A Promise to protect an item.
     */
    function protectItem(requestOptions) {
        var owner = determineOwner(requestOptions);
        var url = arcgisRestRequest.getPortalUrl(requestOptions) + "/content/users/" + owner + "/items/" + requestOptions.id + "/protect";
        return arcgisRestRequest.request(url, requestOptions);
    }
    /**
     * Unprotect an item. See the [REST Documentation](https://developers.arcgis.com/rest/users-groups-and-items/unprotect.htm) for more information.
     *
     * @param requestOptions - Options for the request
     * @returns A Promise to unprotect an item.
     */
    function unprotectItem(requestOptions) {
        var owner = determineOwner(requestOptions);
        var url = arcgisRestRequest.getPortalUrl(requestOptions) + "/content/users/" + owner + "/items/" + requestOptions.id + "/unprotect";
        return arcgisRestRequest.request(url, requestOptions);
    }

    /* Copyright (c) 2018 Environmental Systems Research Institute, Inc.
     * Apache-2.0 */
    /**
     * ```js
     * import { removeItem } from '@esri/arcgis-rest-items';
     * //
     * removeItem({
     *   id: "3ef",
     *   authentication
     * })
     * ```
     * Delete an item from the portal. See the [REST Documentation](https://developers.arcgis.com/rest/users-groups-and-items/delete-item.htm) for more information.
     *
     * @param requestOptions - Options for the request
     * @returns A Promise that deletes an item.
     */
    function removeItem(requestOptions) {
        var owner = determineOwner(requestOptions);
        var url = arcgisRestRequest.getPortalUrl(requestOptions) + "/content/users/" + owner + "/items/" + requestOptions.id + "/delete";
        return arcgisRestRequest.request(url, requestOptions);
    }
    /**
     * ```js
     * import { removeItemRelationship } from '@esri/arcgis-rest-items';
     * //
     * removeItemRelationship({
     *   originItemId: '3ef',
     *   destinationItemId: 'ae7',
     *   relationshipType: 'Service2Layer',
     *   authentication
     * })
     *   .then(response)
     * ```
     * Remove a relationship between two items. See the [REST Documentation](https://developers.arcgis.com/rest/users-groups-and-items/delete-relationship.htm) for more information.
     *
     * @param requestOptions - Options for the request
     * @returns A Promise to add item resources.
     */
    function removeItemRelationship(requestOptions) {
        var owner = determineOwner(requestOptions);
        var url = arcgisRestRequest.getPortalUrl(requestOptions) + "/content/users/" + owner + "/removeRelationship";
        var options = __assign({ params: {} }, requestOptions);
        arcgisRestRequest.appendCustomParams(requestOptions, options);
        return arcgisRestRequest.request(url, options);
    }
    /**
     * Remove a resource associated with an item
     *
     * @param requestOptions - Options for the request
     * @returns A Promise that deletes an item resource.
     */
    function removeItemResource(requestOptions) {
        var owner = determineOwner(requestOptions);
        var url = arcgisRestRequest.getPortalUrl(requestOptions) + "/content/users/" + owner + "/items/" + requestOptions.id + "/removeResources";
        // mix in user supplied params
        requestOptions.params = __assign({}, requestOptions.params, { resource: requestOptions.resource });
        return arcgisRestRequest.request(url, requestOptions);
    }
    /**
     * ```js
     * import { removeFolder } from "@esri/arcgis-rest-items";
     * //
     * removeFolder({
     *   folderId: "fe4",
     *   owner: "c@sey",
     *   authentication
     * })
     *   .then(response)
     *
     * ```
     * Delete a non-root folder and all the items it contains. See the [REST
     * Documentation](https://developers.arcgis.com/rest/users-groups-and-items/delete-folder.htm) for
     * more information.
     *
     * @param requestOptions - Options for the request
     * @returns A Promise that deletes a folder
     */
    function removeFolder(requestOptions) {
        var owner = determineOwner(requestOptions);
        var url = arcgisRestRequest.getPortalUrl(requestOptions) + "/content/users/" + encodeURIComponent(owner) + "/" + requestOptions.folderId + "/delete";
        return arcgisRestRequest.request(url, requestOptions);
    }

    /* Copyright (c) 2018 Environmental Systems Research Institute, Inc.
     * Apache-2.0 */
    /**
     * ```js
     * import { searchItems } from '@esri/arcgis-rest-items';
     * //
     * searchItems('water')
     *   .then(response) // response.total => 355
     * ```
     * Search a portal for items. See the [REST Documentation](https://developers.arcgis.com/rest/users-groups-and-items/search.htm) for more information.
     *
     * @param search - A string or RequestOptions object to pass through to the endpoint.
     * @returns A Promise that will resolve with the data from the response.
     */
    function searchItems(search) {
        var options = {
            httpMethod: "GET",
            params: {}
        };
        if (typeof search === "string") {
            options.params.q = search;
        }
        else {
            // mixin user supplied requestOptions with defaults
            options = __assign({}, options, search);
            // mixin arbitrary request parameters with search form
            options.params = __assign({}, search.params, search.searchForm);
        }
        // construct the search url
        var url = arcgisRestRequest.getPortalUrl(options) + "/search";
        // send the request
        return arcgisRestRequest.request(url, options);
    }

    /* Copyright (c) 2018 Environmental Systems Research Institute, Inc.
     * Apache-2.0 */
    /**
     * ```js
     * import { updateItem } from '@esri/arcgis-rest-items';
     * //
     * updateItem({
     *   item: {
     *     id: "3ef",
     *     description: "A three hour tour"
     *   },
     *   authentication
     * })
     *   .then(response)
     * ```
     * Update an Item. See the [REST Documentation](https://developers.arcgis.com/rest/users-groups-and-items/update-item.htm) for more information.
     *
     * @param item - The item to update.
     * @param requestOptions - Options for the request.
     * @returns A Promise that updates an item.
     */
    function updateItem(requestOptions) {
        var owner = determineOwner(requestOptions);
        var url = arcgisRestRequest.getPortalUrl(requestOptions) + "/content/users/" + owner + "/items/" + requestOptions.item.id + "/update";
        // serialize the item into something Portal will accept
        requestOptions.params = __assign({}, requestOptions.params, serializeItem(requestOptions.item));
        return arcgisRestRequest.request(url, requestOptions);
    }
    /**
     * ```js
     * import { updateItemResource } from '@esri/arcgis-rest-items';
     * //
     * updateItemResource({
     *   id: '3ef',
     *   resource: file,
     *   name: 'bigkahuna.jpg',
     *   authentication
     * })
     *   .then(response)
     * ```
     * Update a resource associated with an item. See the [REST Documentation](https://developers.arcgis.com/rest/users-groups-and-items/update-resources.htm) for more information.
     *
     * @param requestOptions - Options for the request
     * @returns A Promise that updates an item resource.
     */
    function updateItemResource(requestOptions) {
        var owner = determineOwner(requestOptions);
        var url = arcgisRestRequest.getPortalUrl(requestOptions) + "/content/users/" + owner + "/items/" + requestOptions.id + "/updateResources";
        // mix in user supplied params
        requestOptions.params = __assign({ file: requestOptions.resource, fileName: requestOptions.name, text: requestOptions.content }, requestOptions.params);
        // only override whatever access was specified previously if 'private' was passed explicitly
        if (typeof requestOptions.private !== "undefined") {
            requestOptions.params.access = requestOptions.private
                ? "private"
                : "inherit";
        }
        return arcgisRestRequest.request(url, requestOptions);
    }
    /**
     * ```js
     * import { moveItem } from '@esri/arcgis-rest-items';
     * //
     * moveItem({
     *   itemId: "3ef",
     *   folderId: "7c5",
     *   authentication: userSession
     * })
     * ```
     * Move an item to a folder. See the [REST Documentation](https://developers.arcgis.com/rest/users-groups-and-items/move-item.htm) for more information.
     *
     * @param requestOptions - Options for the request
     * @returns A Promise that resolves with owner and folder details once the move has been completed
     */
    function moveItem(requestOptions) {
        var owner = determineOwner(requestOptions);
        var url = arcgisRestRequest.getPortalUrl(requestOptions) + "/content/users/" + owner + "/items/" + requestOptions.itemId + "/move";
        var folderId = requestOptions.folderId;
        if (!folderId) {
            folderId = "/";
        }
        requestOptions.params = __assign({ folder: folderId }, requestOptions.params);
        return arcgisRestRequest.request(url, requestOptions);
    }

    /* Copyright (c) 2018 Environmental Systems Research Institute, Inc.
     * Apache-2.0 */

    exports.addItemJsonData = addItemJsonData;
    exports.addItemData = addItemData;
    exports.addItemRelationship = addItemRelationship;
    exports.addItemResource = addItemResource;
    exports.createFolder = createFolder;
    exports.createItemInFolder = createItemInFolder;
    exports.createItem = createItem;
    exports.getItem = getItem;
    exports.getItemData = getItemData;
    exports.getRelatedItems = getRelatedItems;
    exports.getItemResources = getItemResources;
    exports.getItemGroups = getItemGroups;
    exports.protectItem = protectItem;
    exports.unprotectItem = unprotectItem;
    exports.removeItem = removeItem;
    exports.removeItemRelationship = removeItemRelationship;
    exports.removeItemResource = removeItemResource;
    exports.removeFolder = removeFolder;
    exports.searchItems = searchItems;
    exports.updateItem = updateItem;
    exports.updateItemResource = updateItemResource;
    exports.moveItem = moveItem;
    exports.serializeItem = serializeItem;
    exports.determineOwner = determineOwner;

    Object.defineProperty(exports, '__esModule', { value: true });

})));
