// @esri/arcgis-rest-items/dist/node/get.js
"use strict";
/* Copyright (c) 2018 Environmental Systems Research Institute, Inc.
 * Apache-2.0 */
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var arcgis_rest_request_1 = require("@esri/arcgis-rest-request");
/**
 * ```
 * import { getItem } from "@esri/arcgis-rest-items";
 * //
 * getItem("ae7")
 *   .then(response);
 * // or
 * getItem("ae7", { authentication })
 *   .then(response)
 * ```
 * Get an item by id. See the [REST Documentation](https://developers.arcgis.com/rest/users-groups-and-items/item.htm) for more information.
 *
 * @param id - Item Id
 * @param requestOptions - Options for the request
 * @returns A Promise that will resolve with the data from the response.
 */
function getItem(id, requestOptions) {
    var url = arcgis_rest_request_1.getPortalUrl(requestOptions) + "/content/items/" + id;
    // default to a GET request
    var options = tslib_1.__assign({ httpMethod: "GET" }, requestOptions);
    return arcgis_rest_request_1.request(url, options);
}
exports.getItem = getItem;
/**
 * ```
 * import { getItemData } from "@esri/arcgis-rest-items";
 * //
 * getItemData("ae7")
 *   .then(response)
 * // or
 * getItemData("ae7", { authentication })
 *   .then(response)
 * ```
 * Get the /data for an item. See the [REST Documentation](https://developers.arcgis.com/rest/users-groups-and-items/item-data.htm) for more information.
 * @param id - Item Id
 * @param requestOptions - Options for the request
 * @returns A Promise that will resolve with the json data for the item.
 */
function getItemData(id, requestOptions) {
    var url = arcgis_rest_request_1.getPortalUrl(requestOptions) + "/content/items/" + id + "/data";
    // default to a GET request
    var options = tslib_1.__assign({ httpMethod: "GET", params: {} }, requestOptions);
    if (options.file) {
        options.params.f = null;
    }
    return arcgis_rest_request_1.request(url, options);
}
exports.getItemData = getItemData;
/**
 * ```
 * import { getRelatedItems } from "@esri/arcgis-rest-items";
 * //
 * getRelatedItems({
 *   id: "ae7",
 *   relationshipType: "Service2Layer" // or several ["Service2Layer", "Map2Area"]
 * })
 *   .then(response)
 * ```
 * Get the related items. See the [REST Documentation](https://developers.arcgis.com/rest/users-groups-and-items/related-items.htm) for more information.
 *
 * @param requestOptions - Options for the request
 * @returns A Promise to get some item resources.
 */
function getRelatedItems(requestOptions) {
    var url = arcgis_rest_request_1.getPortalUrl(requestOptions) + "/content/items/" + requestOptions.id + "/relatedItems";
    var options = tslib_1.__assign({ httpMethod: "GET", params: {
            direction: requestOptions.direction
        } }, requestOptions);
    if (typeof requestOptions.relationshipType === "string") {
        options.params.relationshipType = requestOptions.relationshipType;
    }
    else {
        options.params.relationshipTypes = requestOptions.relationshipType;
    }
    delete options.direction;
    delete options.relationshipType;
    return arcgis_rest_request_1.request(url, options);
}
exports.getRelatedItems = getRelatedItems;
/**
 * Get the resources associated with an item
 *
 * @param requestOptions - Options for the request
 * @returns A Promise to get some item resources.
 */
function getItemResources(requestOptions) {
    var url = arcgis_rest_request_1.getPortalUrl(requestOptions) + "/content/items/" + requestOptions.id + "/resources";
    // mix in user supplied params
    requestOptions.params = tslib_1.__assign({}, requestOptions.params, { num: 1000 });
    // at v2, the argument signature of this method should match getItemData() and getItemGroups() if requests can be made anonymously
    return arcgis_rest_request_1.request(url, requestOptions);
}
exports.getItemResources = getItemResources;
/**
 * ```js
 * import { getItemGroups } from "@esri/arcgis-rest-items";
 * //
 * getItemGroups("30e5fe3149c34df1ba922e6f5bbf808f")
 *   .then(response)
 * ```
 * Lists the groups of which the item is a part, only showing the groups that the calling user can access. See the [REST Documentation](https://developers.arcgis.com/rest/users-groups-and-items/groups.htm) for more information.
 *
 * @param id - The Id of the item to query group association for.
 * @param requestOptions - Options for the request
 * @returns A Promise to get some item groups.
 */
function getItemGroups(id, requestOptions) {
    var url = arcgis_rest_request_1.getPortalUrl(requestOptions) + "/content/items/" + id + "/groups";
    return arcgis_rest_request_1.request(url, requestOptions);
}
exports.getItemGroups = getItemGroups;
